Re-use ContentLayers overlay and underlay Elements instead of always re-inflating (Resolves #1236)

matthew-carroll opened the pull request:
Re-use ContentLayers overlay and underlay Elements instead of always re-inflating (Resolves #1236)

knopp commented:
This seems like a good way forward. I haven't looked at the failed tests yet, but I was able to remove `..startBlinking()` from `_CaretDocumentOverlayState.initState()` (which cases repaints even when not actively editing)/

knopp commented:
So the problem here seems to be that you're calling `child.rebuild(force: true);` during `buildLayers`, but the child is a `SingleChildRenderObjectElement` which already has a child and rebuild doesn't do anything useful. It definitely doesn't traverse down to children - that would normally be a performance issue given that Flutter already has finer grained dirty tracking.

Given that it seems that you want to avoid regular dirty element tracking here and just force rebuild of entire subtree, you probably needs bigger hammer than `rebuild`. Something like `reassemble` should do the trick, with the caveat that this is not what it is intended for (and is marked as protected - though calling it from ContentLayersElement the analyzer won't complain).

Changing `child.rebuild(force: true);` into `child.reassemble();` should make the tests pass.

matthew-carroll commented:
Do you have any recommendation for alternative approaches? I'm not married to anything in particular. I just want to make sure that the overlay and underlay layers are given an opportunity to build their subtree after the content is laid out...

knopp commented:
So, without using reassemble (which is quite hacky), you could change `ContentLayers` to take list of build methods instead of widgets. That way you will get new widget instances for every `buildLayer` call and Flutter will update nested elements in the tree.

I tried changing `final List<Widget> overlays` to `final List<WidgetBuilder> overlays` (same for underlays) and it seems to work fine.

knopp commented:
https://github.com/superlistapp/super_editor/pull/1240

knopp commented:
We may need a PR for main to replace `textScaleFactorOf` with `textScalerOf`

matthew-carroll commented:
@knopp I switched the layers to builders. I also updated the caret overlay to only start blinking when there's a non-null selection. Let me know if this PR looks good to you.

knopp commented on super_editor/lib/src/infrastructure/content_layers.dart:
```diff
@@ -132,19 +142,29 @@ class ContentLayersElement extends RenderObjectElement {
 
     owner!.buildScope(this, () {
-      _deactivateLayers();
+      // _deactivateLayers();
```
Remove this?

knopp commented on super_editor/pubspec.yaml:
```diff
@@ -37,13 +37,13 @@ dependencies:
 #  # against changes that they're making to other mono-repo packages
 #  attributed_text:
 #    path: ../attributed_text
-#  super_text_layout:
```
Is this committed on purpose?

matthew-carroll replied:
Yeah - I changed the blink controller API, which is in super_text_layout. I'll need to cut a new release of super_text_layout to be able to remove this override.

knopp reviewed (approved):
Just some nits otherwise LGTM.
